File: mobile/lib/shared/contextual_agent/contextual_agent_conversation_policy.dart
// Contextual agent audience + reply-placement policy (Flutter).
//
// Contract: tests/fixtures/contextual-agent-conversation-cases.json
// Pure resolver plus send-path helpers for unaddressed-channel mode.

enum UnaddressedChannelAgentMode { allChannelAgents, mentionsOnly }

sealed class ReplyPlacement {
  const ReplyPlacement();
}

class TopLevelPlacement extends ReplyPlacement {
  const TopLevelPlacement();
}

class ThreadRootPlacement extends ReplyPlacement {
  const ThreadRootPlacement(this.eventId);
  final String eventId;
}

class UnconstrainedPlacement extends ReplyPlacement {
  const UnconstrainedPlacement();
}

class ContextualAgentConversationInput {
  const ContextualAgentConversationInput({
    required this.conversation,
    required this.messagePosition,
    required this.senderClass,
    required this.unaddressedMode,
    required this.keepAddressedAgentsActive,
    required this.explicitMentionPubkeys,
    required this.currentAgentPubkey,
    required this.channelMemberPubkeys,
    required this.verifiedChannelAgentPubkeys,
    this.unverifiedAgentPubkeys = const [],
    this.nonMemberAgentPubkeys = const [],
    required this.threadRootEventId,
    this.replyingUnderEventId,
    required this.persistentThreadAudience,
    required this.manualRemovedPubkeys,
    required this.recipientLoadError,
    this.humanMessageEventId,
  });

  final String conversation;
  final String messagePosition;
  final String senderClass;
  final UnaddressedChannelAgentMode unaddressedMode;
  final bool keepAddressedAgentsActive;
  final List<String> explicitMentionPubkeys;
  final String? currentAgentPubkey;
  final List<String> channelMemberPubkeys;
  final List<String> verifiedChannelAgentPubkeys;
  final List<String> unverifiedAgentPubkeys;
  final List<String> nonMemberAgentPubkeys;
  final String? threadRootEventId;
  final String? replyingUnderEventId;
  final List<String> persistentThreadAudience;
  final List<String> manualRemovedPubkeys;
  final bool recipientLoadError;
  final String? humanMessageEventId;
}

class ContextualAgentConversationDecision {
  const ContextualAgentConversationDecision({
    required this.audiencePubkeys,
    required this.replyPlacement,
    required this.sharedThread,
    required this.retainDraft,
    this.nestUnderAgentReply = false,
  });

  final List<String> audiencePubkeys;
  final ReplyPlacement replyPlacement;
  final bool sharedThread;
  final bool retainDraft;
  final bool nestUnderAgentReply;
}

String _normalizePubkey(String pubkey) => pubkey.trim().toLowerCase();

List<String> _uniqueSorted(Iterable<String> pubkeys) {
  final set = <String>{};
  for (final pk in pubkeys) {
    final n = _normalizePubkey(pk);
    if (n.isNotEmpty) set.add(n);
  }
  final list = set.toList()..sort();
  return list;
}

Set<String> _eligibleChannelAgents(ContextualAgentConversationInput input) {
  final members = input.channelMemberPubkeys.map(_normalizePubkey).toSet();
  return input.verifiedChannelAgentPubkeys
      .map(_normalizePubkey)
      .where(members.contains)
      .toSet();
}

List<String> _filterToEligible(List<String> candidates, Set<String> eligible) {
  return _uniqueSorted(
    candidates.map(_normalizePubkey).where(eligible.contains),
  );
}

ReplyPlacement _placementFor(
  ContextualAgentConversationInput input,
  int audienceCount,
) {
  if (input.messagePosition == 'in-thread' &&
      input.threadRootEventId != null &&
      input.threadRootEventId!.isNotEmpty) {
    return ThreadRootPlacement(input.threadRootEventId!);
  }
  if (audienceCount >= 2) {
    final eventId = input.humanMessageEventId ?? input.threadRootEventId;
    if (eventId != null && eventId.isNotEmpty) {
      return ThreadRootPlacement(eventId);
    }
  }
  return const TopLevelPlacement();
}

/// Resolve audience and reply placement for a human/agent send path.
ContextualAgentConversationDecision resolveContextualAgentConversation(
  ContextualAgentConversationInput input,
) {
  if (input.recipientLoadError) {
    return const ContextualAgentConversationDecision(
      audiencePubkeys: [],
      replyPlacement: TopLevelPlacement(),
      sharedThread: false,
      retainDraft: true,
    );
  }

  if (input.senderClass == 'agent') {
    return const ContextualAgentConversationDecision(
      audiencePubkeys: [],
      replyPlacement: UnconstrainedPlacement(),
      sharedThread: false,
      retainDraft: false,
    );
  }

  if (input.conversation == 'direct') {
    final current = input.currentAgentPubkey;
    final audience = current == null || current.isEmpty
        ? <String>[]
        : [_normalizePubkey(current)];
    return ContextualAgentConversationDecision(
      audiencePubkeys: audience,
      replyPlacement: _placementFor(input, audience.length),
      sharedThread: false,
      retainDraft: false,
    );
  }

  final eligible = _eligibleChannelAgents(input);
  final removed = input.manualRemovedPubkeys.map(_normalizePubkey).toSet();

  final explicit = _filterToEligible(
    input.explicitMentionPubkeys,
    eligible,
  ).where((pk) => !removed.contains(pk)).toList();

  late final List<String> audience;
  if (explicit.isNotEmpty) {
    audience = explicit;
  } else {
    final persistent = input.keepAddressedAgentsActive
        ? _filterToEligible(
            input.persistentThreadAudience,
            eligible,
          ).where((pk) => !removed.contains(pk)).toList()
        : <String>[];
    if (persistent.isNotEmpty) {
      audience = persistent;
    } else if (input.unaddressedMode ==
        UnaddressedChannelAgentMode.allChannelAgents) {
      audience = _uniqueSorted(
        eligible,
      ).where((pk) => !removed.contains(pk)).toList();
    } else {
      audience = [];
    }
  }

  return ContextualAgentConversationDecision(
    audiencePubkeys: audience,
    replyPlacement: _placementFor(input, audience.length),
    sharedThread: audience.length >= 2,
    retainDraft: false,
  );
}

File: desktop/src/features/channels/lib/contextualAgentConversationPolicy.ts
/**
 * Contextual agent audience + reply-placement policy (Desktop).
 *
 * Contract: tests/fixtures/contextual-agent-conversation-cases.json
 *
 * Pure resolver — no I/O. Device-local mode persistence lives in
 * `unaddressedChannelAgentMode.ts`.
 */

import { normalizePubkey } from "@/shared/lib/pubkey.ts";

export type UnaddressedChannelAgentMode =
  | "all-channel-agents"
  | "mentions-only";

export type ReplyPlacement =
  | { kind: "top-level" }
  | { kind: "thread-root"; eventId: string }
  | { kind: "unconstrained" };

export type ContextualAgentConversationInput = {
  conversation: "direct" | "channel";
  messagePosition: "top-level" | "in-thread";
  senderClass: "human" | "agent";
  unaddressedMode: UnaddressedChannelAgentMode;
  keepAddressedAgentsActive: boolean;
  explicitMentionPubkeys: string[];
  currentAgentPubkey: string | null;
  channelMemberPubkeys: string[];
  verifiedChannelAgentPubkeys: string[];
  unverifiedAgentPubkeys?: string[];
  nonMemberAgentPubkeys?: string[];
  threadRootEventId: string | null;
  replyingUnderEventId?: string | null;
  persistentThreadAudience: string[];
  manualRemovedPubkeys: string[];
  recipientLoadError: boolean;
  /** Id of the human message that becomes a multi-agent thread root. */
  humanMessageEventId?: string | null;
};

export type ContextualAgentConversationDecision = {
  audiencePubkeys: string[];
  replyPlacement: ReplyPlacement;
  sharedThread: boolean;
  retainDraft: boolean;
  nestUnderAgentReply?: boolean;
};

function uniqueSorted(pubkeys: Iterable<string>): string[] {
  const set = new Set<string>();
  for (const pk of pubkeys) {
    const n = normalizePubkey(pk);
    if (n) set.add(n);
  }
  return [...set].sort();
}

function asSet(pubkeys: readonly string[]): Set<string> {
  return new Set(pubkeys.map(normalizePubkey).filter(Boolean));
}

/**
 * Verified current-channel agents only: intersection of membership and
 * verified agent evidence. Never community/relay-wide fanout.
 */
function eligibleChannelAgents(
  input: ContextualAgentConversationInput,
): Set<string> {
  const members = asSet(input.channelMemberPubkeys);
  const verified = asSet(input.verifiedChannelAgentPubkeys);
  const eligible = new Set<string>();
  for (const pk of verified) {
    if (members.has(pk)) eligible.add(pk);
  }
  return eligible;
}

function filterToEligible(
  candidates: readonly string[],
  eligible: Set<string>,
): string[] {
  return uniqueSorted(
    candidates.filter((pk) => eligible.has(normalizePubkey(pk))),
  );
}

/**
 * Resolve audience and reply placement for a human/agent send path.
 *
 * Precedence:
 * 1. Recipient-load errors fail closed (retain draft).
 * 2. Agent-authored traffic is unconstrained.
 * 3. Explicit @mentions (eligible agents only).
 * 4. Persistent audience when Keep addressed agents active (minus manual removals).
 * 5. Unaddressed mode: all verified channel agents, or mentions-only (none).
 * 6. Direct conversations always resolve to the current agent.
 */
export function resolveContextualAgentConversation(
  input: ContextualAgentConversationInput,
): ContextualAgentConversationDecision {
  if (input.recipientLoadError) {
    return {
      audiencePubkeys: [],
      replyPlacement: { kind: "top-level" },
      sharedThread: false,
      retainDraft: true,
      nestUnderAgentReply: false,
    };
  }

  if (input.senderClass === "agent") {
    return {
      audiencePubkeys: [],
      replyPlacement: { kind: "unconstrained" },
      sharedThread: false,
      retainDraft: false,
      nestUnderAgentReply: false,
    };
  }

  // Human path
  if (input.conversation === "direct") {
    const current = input.currentAgentPubkey
      ? normalizePubkey(input.currentAgentPubkey)
      : null;
    const audience = current ? [current] : [];
    return {
      audiencePubkeys: audience,
      replyPlacement: placementFor(input, audience.length),
      sharedThread: false,
      retainDraft: false,
      nestUnderAgentReply: false,
    };
  }

  // Channel path
  const eligible = eligibleChannelAgents(input);
  const removed = asSet(input.manualRemovedPubkeys);
  const explicit = filterToEligible(
    input.explicitMentionPubkeys,
    eligible,
  ).filter((pk) => !removed.has(pk));

  let audience: string[];

  if (explicit.length > 0) {
    audience = explicit;
  } else {
    const persistent = input.keepAddressedAgentsActive
      ? filterToEligible(input.persistentThreadAudience, eligible).filter(
          (pk) => !removed.has(pk),
        )
      : [];

    if (persistent.length > 0) {
      audience = persistent;
    } else if (input.unaddressedMode === "all-channel-agents") {
      audience = uniqueSorted(eligible).filter((pk) => !removed.has(pk));
    } else {
      // mentions-only, no explicit, no persistent
      audience = [];
    }
  }

  const sharedThread = audience.length >= 2;
  return {
    audiencePubkeys: audience,
    replyPlacement: placementFor(input, audience.length),
    sharedThread,
    retainDraft: false,
    nestUnderAgentReply: false,
  };
}

function placementFor(
  input: ContextualAgentConversationInput,
  audienceCount: number,
): ReplyPlacement {
  // Already in a thread: always continue at the existing root (never nest under
  // an agent reply).
  if (input.messagePosition === "in-thread" && input.threadRootEventId) {
    return {
      kind: "thread-root",
      eventId: input.threadRootEventId,
    };
  }

  // Multi-agent top-level human message creates one shared thread at the human
  // event.
  if (audienceCount >= 2) {
    const eventId = input.humanMessageEventId ?? input.threadRootEventId;
    if (eventId) {
      return { kind: "thread-root", eventId };
    }
  }

  return { kind: "top-level" };
}
